### ISSUE TITLE
multiClassSummary of binary classification errs (unsurpringly but quite harsly) 

### ISSUE THREAD

**Issue by hadjipantelis** [topepo/caret#587]
Hello,

Using `multiClassSummary` for a binary classifier is probably a bit redundant but `train` fails extremely inelegantly and uninformatively when this happens. 

```R
rm(list=ls())
devtools::load_all('/home/pantelishadjipa/git_projects/caret/pkg/caret/')
N = 1010;
M = 10;
set.seed(1)
xTrain = matrix( runif(N*M), nrow = N);
colnames(xTrain) <- 
  sapply(1:M, function(u) paste0(collapse = '', letters[sample(26, 3, replace = TRUE)] ) )
yTrain = as.factor( letters[sample(2, N, replace = TRUE)])

trCntlListMulti  <- trainControl(method = "repeatedcv", verboseIter =TRUE, 
                                 classProbs = TRUE, 
                                 summaryFunction = multiClassSummary ) 
 
enFitMulti <- train( x = xTrain, y = yTrain, trControl = trCntlListMulti,  
                     method = "glmnet", tuneLength = 4 ) #  Errs

# Error in vector(type, length) : 
# vector: cannot make a vector of mode 'NULL'.
```
It would be nice to either disallow `multiClassSummary` in the case of binary data or at least give some sort of more informative error. (As a somewhat more far reaching feature maybe `multiClassSummary` should automatically work for binary data)

All best, 
Pantelis  

```R
> sessionInfo()
R version 3.3.1 (2016-06-21)
Platform: x86_64-redhat-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C               LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
 [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8    LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] caret_6.0-75    plyr_1.8.4      xgboost_0.6-4   DATB_0.1.0      glmnet_2.0-5    foreach_1.4.3   Matrix_1.2-6    ggplot2_2.2.0  
[9] lattice_0.20-33

loaded via a namespace (and not attached):
 [1] gbm_2.1.1           reshape2_1.4.2      splines_3.3.1       colorspace_1.2-6    testthat_1.0.2      stats4_3.3.1       
 [7] mgcv_1.8-12         chron_2.3-47        survival_2.39-4     ModelMetrics_1.1.0  e1071_1.6-7         nloptr_1.0.4       
[13] withr_1.0.2         DBI_0.5-1           stringr_1.1.0       MatrixModels_0.4-1  munsell_0.4.3       gtable_0.2.0       
[19] moments_0.14        devtools_1.12.0     codetools_0.2-14    memoise_1.0.0       SparseM_1.72        quantreg_5.29      
[25] pbkrtest_0.4-6      parallel_3.3.1      class_7.3-14        Rcpp_0.12.8         scales_0.4.1        lme4_1.1-12        
[31] digest_0.6.10       stringi_1.1.1       dplyr_0.5.0         grid_3.3.1          tools_3.3.1         magrittr_1.5       
[37] lazyeval_0.2.0      tibble_1.2          randomForest_4.6-12 crayon_1.3.2        car_2.1-4           MASS_7.3-45        
[43] RODBC_1.3-13        data.table_1.9.6    assertthat_0.1      minqa_1.2.4         roxygen2_5.0.1      iterators_1.0.8    
[49] R6_2.2.0            nnet_7.3-12         nlme_3.1-128        compiler_3.3.1   
```

**Comment by RobertKomara** [comment]
This was supposed to have been fixed re: issues #564 #528 etc. but still seems to be a problem:

I installed a fresh copy of 6.0-75 yesterday and ran the following tests. The last version that seems to be working is 6.0-71:

Caret Test
================

Running Caret 6.0-71
--------------------

``` r
library(caret)
```

    ## Loading required package: lattice

    ## Loading required package: ggplot2

``` r
set.seed(1)
dat <- twoClassSim(100)
X <- dat[,1:5]
y <- dat[["Class"]]

model_class <- train(
  X, y, method='xgbTree',
  metric='Kappa',
  trControl=trainControl(
    method="cv", 
    number=5,
    classProbs=TRUE, 
    summaryFunction=multiClassSummary)
)
```

    ## Loading required package: xgboost

    ## Loading required package: plyr

``` r
print(model_class)
```

    ## eXtreme Gradient Boosting 
    ## 
    ## 100 samples
    ##   5 predictor
    ##   2 classes: 'Class1', 'Class2' 
    ## 
    ## No pre-processing
    ## Resampling: Cross-Validated (5 fold) 
    ## Summary of sample sizes: 80, 80, 81, 80, 79 
    ## Resampling results across tuning parameters:
    ## 
    ##   eta  max_depth  colsample_bytree  nrounds  logLoss    ROC      
    ##   0.3  1          0.6                50      0.4687844  0.8435113
    ##   0.3  1          0.6               100      0.4919091  0.8353480
    ##   0.3  1          0.6               150      0.5061484  0.8285453
    ##   0.3  1          0.8                50      0.4699971  0.8273940
    ##   0.3  1          0.8               100      0.4815070  0.8377289
    ##   0.3  1          0.8               150      0.4927652  0.8296442
    ##   0.3  2          0.6                50      0.5146657  0.8222920
    ##   0.3  2          0.6               100      0.5500785  0.8196756
    ##   0.3  2          0.6               150      0.5853222  0.8037153
    ##   0.3  2          0.8                50      0.4685916  0.8464155
    ##   0.3  2          0.8               100      0.4993035  0.8403977
    ##   0.3  2          0.8               150      0.4957119  0.8471481
    ##   0.3  3          0.6                50      0.5344720  0.8082679
    ##   0.3  3          0.6               100      0.5715378  0.8142334
    ##   0.3  3          0.6               150      0.5851659  0.8255887
    ##   0.3  3          0.8                50      0.5058371  0.8207745
    ##   0.3  3          0.8               100      0.5388538  0.8146520
    ##   0.3  3          0.8               150      0.5548595  0.8091575
    ##   0.4  1          0.6                50      0.4754303  0.8252486
    ##   0.4  1          0.6               100      0.5015120  0.8274464
    ##   0.4  1          0.6               150      0.5146291  0.8223182
    ##   0.4  1          0.8                50      0.4828160  0.8287284
    ##   0.4  1          0.8               100      0.4898309  0.8208006
    ##   0.4  1          0.8               150      0.5022138  0.8186028
    ##   0.4  2          0.6                50      0.5085914  0.8389325
    ##   0.4  2          0.6               100      0.5578814  0.8301936
    ##   0.4  2          0.6               150      0.5749214  0.8344322
    ##   0.4  2          0.8                50      0.5150736  0.8221612
    ##   0.4  2          0.8               100      0.5750871  0.8028257
    ##   0.4  2          0.8               150      0.5904701  0.8020931
    ##   0.4  3          0.6                50      0.5141991  0.8121402
    ##   0.4  3          0.6               100      0.5798702  0.8106750
    ##   0.4  3          0.6               150      0.6018146  0.8086342
    ##   0.4  3          0.8                50      0.5236716  0.8151753
    ##   0.4  3          0.8               100      0.5289809  0.8391941
    ##   0.4  3          0.8               150      0.5319494  0.8465201
    ##   Accuracy   Kappa      Sensitivity  Specificity  Pos_Pred_Value
    ##   0.7407519  0.4195718  0.8043956    0.6142857    0.8054212     
    ##   0.7707519  0.4947856  0.8043956    0.7000000    0.8456643     
    ##   0.7507519  0.4536353  0.7890110    0.6714286    0.8277156     
    ##   0.7302757  0.3814940  0.8186813    0.5571429    0.7856410     
    ##   0.7507519  0.4434805  0.8043956    0.6428571    0.8205728     
    ##   0.7507519  0.4434805  0.8043956    0.6428571    0.8205728     
    ##   0.7507519  0.4453917  0.8043956    0.6428571    0.8176923     
    ##   0.7602757  0.4573044  0.8186813    0.6428571    0.8244689     
    ##   0.7702757  0.4771782  0.8340659    0.6428571    0.8257509     
    ##   0.7808020  0.4950716  0.8494505    0.6428571    0.8343823     
    ##   0.7712782  0.4901606  0.8197802    0.6714286    0.8371753     
    ##   0.7708020  0.4775680  0.8340659    0.6428571    0.8302156     
    ##   0.7597995  0.4538575  0.8329670    0.6142857    0.8101832     
    ##   0.7597995  0.4555951  0.8329670    0.6142857    0.8083516     
    ##   0.7897995  0.5192814  0.8637363    0.6428571    0.8259341     
    ##   0.7602757  0.4715031  0.8032967    0.6714286    0.8287646     
    ##   0.7602757  0.4701870  0.8032967    0.6714286    0.8302797     
    ##   0.7602757  0.4701870  0.8032967    0.6714286    0.8302797     
    ##   0.7502757  0.4343080  0.8186813    0.6142857    0.8110490     
    ##   0.7407519  0.4297265  0.7890110    0.6428571    0.8125641     
    ##   0.7508020  0.4515477  0.8032967    0.6428571    0.8145979     
    ##   0.7407519  0.4195718  0.8043956    0.6142857    0.8054212     
    ##   0.7407519  0.4195718  0.8043956    0.6142857    0.8054212     
    ##   0.7407519  0.4195718  0.8043956    0.6142857    0.8054212     
    ##   0.7808020  0.5007180  0.8494505    0.6428571    0.8283974     
    ##   0.7903258  0.5146565  0.8791209    0.6142857    0.8174451     
    ##   0.8003258  0.5387388  0.8791209    0.6428571    0.8315476     
    ##   0.7697995  0.4717719  0.8483516    0.6142857    0.8139927     
    ##   0.7697995  0.4796774  0.8329670    0.6428571    0.8224542     
    ##   0.7297995  0.3735037  0.8329670    0.5285714    0.7757509     
    ##   0.7902757  0.5381386  0.8186813    0.7285714    0.8595338     
    ##   0.7602757  0.4701870  0.8032967    0.6714286    0.8302797     
    ##   0.7702757  0.4888314  0.8186813    0.6714286    0.8317949     
    ##   0.7803258  0.5022752  0.8483516    0.6428571    0.8249542     
    ##   0.7703258  0.4836308  0.8329670    0.6428571    0.8234391     
    ##   0.7703258  0.4836308  0.8329670    0.6428571    0.8234391     
    ##   Neg_Pred_Value  Detection_Rate  Balanced_Accuracy
    ##   0.6289683       0.5315539       0.7093407        
    ##   0.6587302       0.5315539       0.7521978        
    ##   0.6373016       0.5215539       0.7302198        
    ##   0.6121429       0.5410777       0.6879121        
    ##   0.6373016       0.5315539       0.7236264        
    ##   0.6373016       0.5315539       0.7236264        
    ##   0.6396825       0.5315539       0.7236264        
    ##   0.6416667       0.5410777       0.7307692        
    ##   0.6630952       0.5510777       0.7384615        
    ##   0.7083333       0.5616040       0.7461538        
    ##   0.6944444       0.5420802       0.7456044        
    ##   0.6883333       0.5516040       0.7384615        
    ##   0.6607143       0.5506015       0.7236264        
    ##   0.6642857       0.5506015       0.7236264        
    ##   0.7166667       0.5706015       0.7532967        
    ##   0.6539683       0.5310777       0.7373626        
    ##   0.6511905       0.5310777       0.7373626        
    ##   0.6511905       0.5310777       0.7373626        
    ##   0.6416667       0.5410777       0.7164835        
    ##   0.6289683       0.5215539       0.7159341        
    ##   0.6789683       0.5316040       0.7230769        
    ##   0.6289683       0.5315539       0.7093407        
    ##   0.6289683       0.5315539       0.7093407        
    ##   0.6289683       0.5315539       0.7093407        
    ##   0.7200000       0.5616040       0.7461538        
    ##   0.7523810       0.5811278       0.7467033        
    ##   0.7595238       0.5811278       0.7609890        
    ##   0.6807143       0.5606015       0.7313187        
    ##   0.6714286       0.5506015       0.7379121        
    ##   0.6204762       0.5506015       0.6807692        
    ##   0.6861111       0.5410777       0.7736264        
    ##   0.6511905       0.5310777       0.7373626        
    ##   0.6678571       0.5410777       0.7450549        
    ##   0.7214286       0.5611278       0.7456044        
    ##   0.7047619       0.5511278       0.7379121        
    ##   0.7047619       0.5511278       0.7379121        
    ## 
    ## Tuning parameter 'gamma' was held constant at a value of 0
    ## 
    ## Tuning parameter 'min_child_weight' was held constant at a value of 1
    ## Kappa was used to select the optimal model using  the largest value.
    ## The final values used for the model were nrounds = 150, max_depth = 2,
    ##  eta = 0.4, gamma = 0, colsample_bytree = 0.6 and min_child_weight = 1.

``` r
sessionInfo()
```

    ## R version 3.3.2 (2016-10-31)
    ## Platform: x86_64-w64-mingw32/x64 (64-bit)
    ## Running under: Windows 10 x64 (build 10586)
    ## 
    ## locale:
    ## [1] LC_COLLATE=English_United Kingdom.1252 
    ## [2] LC_CTYPE=English_United Kingdom.1252   
    ## [3] LC_MONETARY=English_United Kingdom.1252
    ## [4] LC_NUMERIC=C                           
    ## [5] LC_TIME=English_United Kingdom.1252    
    ## 
    ## attached base packages:
    ## [1] stats     graphics  grDevices utils     datasets  methods   base     
    ## 
    ## other attached packages:
    ## [1] plyr_1.8.4           xgboost_0.4-4        caret_6.0-71        
    ## [4] ggplot2_2.1.0        lattice_0.20-34      RevoUtilsMath_10.0.0
    ## 
    ## loaded via a namespace (and not attached):
    ##  [1] Rcpp_0.12.7        compiler_3.3.2     formatR_1.4       
    ##  [4] nloptr_1.0.4       class_7.3-14       iterators_1.0.8   
    ##  [7] tools_3.3.2        digest_0.6.10      lme4_1.1-12       
    ## [10] evaluate_0.10      gtable_0.2.0       nlme_3.1-128      
    ## [13] mgcv_1.8-15        Matrix_1.2-7.1     foreach_1.4.3     
    ## [16] yaml_2.1.13        parallel_3.3.2     SparseM_1.72      
    ## [19] e1071_1.6-7        stringr_1.1.0      knitr_1.14        
    ## [22] pROC_1.8           RevoUtils_10.0.2   MatrixModels_0.4-1
    ## [25] stats4_3.3.2       rprojroot_1.1      grid_3.3.2        
    ## [28] nnet_7.3-12        data.table_1.9.6   rmarkdown_1.3     
    ## [31] minqa_1.2.4        reshape2_1.4.2     car_2.1-3         
    ## [34] magrittr_1.5       backports_1.0.4    scales_0.4.0      
    ## [37] codetools_0.2-15   htmltools_0.3.5    MASS_7.3-45       
    ## [40] splines_3.3.2      pbkrtest_0.4-6     colorspace_1.2-7  
    ## [43] quantreg_5.29      stringi_1.1.2      munsell_0.4.3     
    ## [46] chron_2.3-47

While 75 is not working:

Caret Test
================

Running Caret 6.0-75
--------------------

``` r
library(caret)
```

    ## Loading required package: lattice

    ## Loading required package: ggplot2

``` r
set.seed(1)
dat <- twoClassSim(100)
X <- dat[,1:5]
y <- dat[["Class"]]

model_class <- train(
  X, y, method='xgbTree',
  metric='Kappa',
  trControl=trainControl(
    method="cv", 
    number=5,
    classProbs=TRUE, 
    summaryFunction=multiClassSummary)
)
```

    ## Loading required package: xgboost

    ## Loading required package: plyr

    ## Error in vector(type, length): vector: cannot make a vector of mode 'NULL'.

``` r
print(model_class)
```

    ## Error in print(model_class): object 'model_class' not found

``` r
sessionInfo()
```

    ## R version 3.3.2 (2016-10-31)
    ## Platform: x86_64-w64-mingw32/x64 (64-bit)
    ## Running under: Windows 10 x64 (build 10586)
    ## 
    ## locale:
    ## [1] LC_COLLATE=English_United Kingdom.1252 
    ## [2] LC_CTYPE=English_United Kingdom.1252   
    ## [3] LC_MONETARY=English_United Kingdom.1252
    ## [4] LC_NUMERIC=C                           
    ## [5] LC_TIME=English_United Kingdom.1252    
    ## 
    ## attached base packages:
    ## [1] stats     graphics  grDevices utils     datasets  methods   base     
    ## 
    ## other attached packages:
    ## [1] plyr_1.8.4           xgboost_0.4-4        caret_6.0-75        
    ## [4] ggplot2_2.1.0        lattice_0.20-34      RevoUtilsMath_10.0.0
    ## 
    ## loaded via a namespace (and not attached):
    ##  [1] Rcpp_0.12.7        compiler_3.3.2     formatR_1.4       
    ##  [4] nloptr_1.0.4       class_7.3-14       iterators_1.0.8   
    ##  [7] tools_3.3.2        digest_0.6.10      lme4_1.1-12       
    ## [10] evaluate_0.10      nlme_3.1-128       gtable_0.2.0      
    ## [13] mgcv_1.8-15        Matrix_1.2-7.1     foreach_1.4.3     
    ## [16] yaml_2.1.13        parallel_3.3.2     SparseM_1.72      
    ## [19] e1071_1.6-7        stringr_1.1.0      knitr_1.14        
    ## [22] RevoUtils_10.0.2   MatrixModels_0.4-1 stats4_3.3.2      
    ## [25] rprojroot_1.1      grid_3.3.2         nnet_7.3-12       
    ## [28] data.table_1.9.6   rmarkdown_1.3      minqa_1.2.4       
    ## [31] reshape2_1.4.2     car_2.1-3          magrittr_1.5      
    ## [34] backports_1.0.4    scales_0.4.0       codetools_0.2-15  
    ## [37] ModelMetrics_1.1.0 htmltools_0.3.5    MASS_7.3-45       
    ## [40] splines_3.3.2      pbkrtest_0.4-6     colorspace_1.2-7  
    ## [43] quantreg_5.29      stringi_1.1.2      munsell_0.4.3     
    ## [46] chron_2.3-47

I thought this may have something to do with using Microsoft's Open R, but the same thing happens with R 3.3.2. I don't know what else to try besides that. Thanks for any help you might be able to provide.

**Comment by hadjipantelis** [comment]
Hello @RobertKomara at least the issue number #528 is not directly related to this. 

In any case I have made a crude fix in my local fork of `caret`; you can find it [here](https://github.com/hadjipantelis/caret). In the fix, if `multiClassSummary` is called in case where one has only two levels/classes it throws a warning about it and then calls `twoClassSummary`. Let me stress that this is just a quick fix I made. I will make a pull request if Max thinks it's alright.

The new code effectively is: 
```R
multiClassSummary <- function (data, lev = NULL, model = NULL){
   ...
   has_class_probs <- all(lev %in% colnames(data))
   if( nlevels(data[, "pred"]) == 2 ){
    warning("declared 'multiClassSummary' in binary classification task") 
    return( twoClassSummary( data, lev = lev, model = model ) )
   }
   if(has_class_probs) {
   ...
```

**Comment by RobertKomara** [comment]
Hi @hadjipantelis, I appreciate the workaround, but I could just call <code>twoClassSummary</code> in the first place. I use a custom measure consisting of **Kappa** and **logLoss** and that is why I use <code>MultiClassSummary</code> in my binary classification task to get those statistics.

**Comment by hadjipantelis** [comment]
@RobertKomara, for that user-case (and irrelevantly to this thread) I fail to see why calling `defaultSummary` and `mnLogLoss` in one go, would not solve your problem. Just write something like:
```R
zongFunc <- function(...){ c(defaultSummary(...), mnLogLoss(...)) }
```
and use `zongFunction` as the `summaryFunction` of `trainControl`. This will give you Kappa and logLoss directly.

**Comment by RobertKomara** [comment]
Hi @hadjipantelis, that would solve my problem, but in general even if one called <code>default</code>, <code>mnLogLoss</code>, and <code>twoClassSummary</code> and combined them all in a wrapper you still would be missing the statistics: **Pos_Pred_Value**, **Neg_Pred_Value**, **Detection_Rate**, and **Balanced_Accuracy** that are available in <code>MultiClassSummary</code> and those are nice to have available. If the idea is to not use <code>MultiClassSummary</code> for binary tasks moving forward, then it would be great if those missing statistics were added to <code>twoClassSummary</code> instead of just losing that functionality. I am a noob here, so if I am missing something obvious, please let me know.

**Comment by achekroud** [comment]
seconding @RobertKomara here -- I was also using `multiClassSummary` as a cheap way of getting detailed performance metrics and also knowing that I was using code that was maintained professionally (I have loved using caret the last few years!). Back in mid November I had some code for a two class problem that used  `multiClassSummary` and ran without errors. Now, using the devtools version of `caret`, the same code fails:
```
# Error in vector(type, length) : 
#   vector: cannot make a vector of mode 'NULL'.
```
since I used the devtools version it seems the logLoss fix in #528 didn't dodge the issue of NAs in 
`resamples <- rbind.fill(result[names(result) == "resamples"])`

instead, incase anyone else is in the same boat, i reverted to some hacky old code I had as follows:

```
## function to get pos and neg pred values
predictiveValues <- function (data, lev = NULL, model = NULL,...){
  PPVobj <- posPredValue(data[, "pred"], data[, "obs"])
  NPVobj <- negPredValue(data[, "pred"], data[, "obs"])
  out <- c(PPVobj, NPVobj)
  names(out) <- c("Pos Pred Value", "Neg Pred Value")
  out}

## function to combine the defaultSummary and the twoClass summary and append the PPV/NPV
custom.summary  <- function(data, lev = NULL, model = NULL, positive = "ytmt",...){
  a <- defaultSummary(data, lev, model)
  b <- twoClassSummary(data, lev, model)
  c <- predictiveValues(data, lev, model)
  out<-c(a,b,c)
  out}
```

This then returns performance metrics that are pretty close to what we used to get from `multiClassSummary`, and although the code is obviously hacky it does return accuracy, kappa, ROC, sensitivity, specificity, ppv, and npv fairly neatly





```
> sessionInfo()
R version 3.3.2 (2016-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] glmnet_2.0-5         Matrix_1.2-7.1       stringr_1.0.0        Hmisc_4.0-0         
 [5] Formula_1.2-1        survival_2.40-1      dplyr_0.5.0          doMC_1.3.4          
 [9] iterators_1.0.8      foreach_1.4.3        e1071_1.6-7          randomForest_4.6-12 
[13] caret_6.0-75         ggplot2_2.2.0        RevoUtilsMath_10.0.0 RevoUtils_10.0.2    
[17] RevoMods_10.0.0      RevoScaleR_9.0.1     lattice_0.20-34      rpart_4.1-10        

loaded via a namespace (and not attached):
 [1] reshape2_1.4.2         splines_3.3.2          colorspace_1.3-1       htmltools_0.3.5       
 [5] stats4_3.3.2           CompatibilityAPI_1.1.0 mgcv_1.8-15            chron_2.3-47          
 [9] ModelMetrics_1.1.0     nloptr_1.0.4           foreign_0.8-67         DBI_0.5-1             
[13] RColorBrewer_1.1-2     plyr_1.8.4             MatrixModels_0.4-1     munsell_0.4.3         
[17] gtable_0.2.0           codetools_0.2-15       knitr_1.15.1           latticeExtra_0.6-28   
[21] SparseM_1.74           mrupdate_1.0.0         quantreg_5.29          pbkrtest_0.4-6        
[25] class_7.3-14           htmlTable_1.7          Rcpp_0.12.7            acepack_1.4.1         
[29] scales_0.4.1           jsonlite_1.1           lme4_1.1-12            gridExtra_2.2.1       
[33] digest_0.6.10          stringi_1.1.1          grid_3.3.2             tools_3.3.2           
[37] magrittr_1.5           lazyeval_0.2.0         tibble_1.2             cluster_2.0.5         
[41] car_2.1-4              MASS_7.3-45            data.table_1.9.6       assertthat_0.1        
[45] minqa_1.2.4            R6_2.2.0               compiler_3.3.2         nnet_7.3-12           
[49] nlme_3.1-128          
```

**Comment by hadjipantelis** [comment]
@achekroud : I think it will be best to open a separate issue. The issue I address is the unexpected error. The issue you and @RobertKomara want to address is that you cannot get multiple metrics in one go. The issue I raised it is implementation-related (it needs code fixing / exception-handling), the issue you raise is design-related (it needs new functionality; you have a working custom summary function, emulating `multiClassSummary` in binary classif. tasks, but you dislike/feel uncomfortable coding it yourself and you would like `caret` to have something like that natively).

**Comment by RobertKomara** [comment]
Hi @hadjipantelis, I think I see the source of the confusion now. We are not requesting new functionality. We are pointing out that previous functionality was broken with release 6.0-72. Namely that <code>multiClassSummary</code> worked for binary cases. You mentioned in the first place that: "It would be nice to either disallow <code>multiClassSummary</code> in the case of binary data or at least give some sort of more informative error. (As a somewhat more far reaching feature maybe <code>multiClassSummary</code> should automatically work for binary data)". We are trying to tell you that <code>multiClassSummary</code> used to automatically work for binary data prior to release 6.0-72 and it would be great if that feature were restored. Please see my example.

**Comment by hadjipantelis** [comment]
Thanks @JackStat I will close this issue.

**Comment by achekroud** [comment]
@jackstat fix #596 seems to work for me (i installed his fork using devtools and multiClassSummary seems to be behaving as it used to before 6.0-72 when you use it in a 2 class problem!
